File: engine/inference_v2/experience_phase1_segment/entry_slice_heads.py
"""Resolve which B-ENTRY lines start a new job entry block for phrase segmentation."""

from __future__ import annotations

from collections import defaultdict

from inference_v2.experience_phase2_divider.date_patterns import has_date_anchor, is_date_token
from inference_v2.experience_phase2_divider.entry_postprocess import (
    _first_meaningful_token,
    _line_text,
    is_bullet_line,
)

MAIN_ENTRY_BULLETS = frozenset({"•", "●", "▪"})
SUB_ENTRY_BULLETS = frozenset({"◦", "∗", "*", "·", "\uf0b7", "\uf0a7"})
STRUCTURAL_PREFIX = frozenset({'"', "'", ",", "|", "(", ")"})


def _line_preds(tokens: list[dict]) -> list[str]:
    return [t.get("bioLabel") or t.get("bio_label") or "O" for t in tokens]


def _first_meaningful_pred(line_toks: list[dict]) -> str | None:
    for t, pred in zip(line_toks, _line_preds(line_toks)):
        tok = (t.get("token") or "").strip()
        if tok and tok not in STRUCTURAL_PREFIX:
            return pred
    return None


def _is_date_only_entry_line(line_toks: list[dict]) -> bool:
    text = _line_text(line_toks)
    first = _first_meaningful_token(line_toks)
    if not first:
        return False
    if first in {"(", "|"} and has_date_anchor(text):
        return True
    if has_date_anchor(text) and not any(
        len((t.get("token") or "").strip()) > 3
        and not is_date_token(t.get("token", ""))
        and (t.get("token") or "").strip() not in SUB_ENTRY_BULLETS | MAIN_ENTRY_BULLETS
        for t in line_toks
    ):
        return True
    return False


def _is_sub_bullet_line(line_toks: list[dict]) -> bool:
    first = _first_meaningful_token(line_toks)
    return first in SUB_ENTRY_BULLETS


def _line_has_b_entry(line_toks: list[dict]) -> bool:
    return any(pred == "B-ENTRY" for pred in _line_preds(line_toks))


def _is_company_or_date_continuation(line_toks: list[dict]) -> bool:
    """Company suffix or parenthetical date line under a job title (not a new entry)."""
    first = _first_meaningful_token(line_toks)
    if not first or first in MAIN_ENTRY_BULLETS:
        return False
    if first in {"(", "|"}:
        return True
    text = _line_text(line_toks).lower()
    if has_date_anchor(_line_text(line_toks)):
        return True
    if any(tok in text for tok in ("pvt", "ltd", "inc", "llc", "limited")):
        return True
    return False


def _is_primary_title_line(line_toks: list[dict]) -> bool:
    first = _first_meaningful_token(line_toks)
    if not first:
        return False
    if first in MAIN_ENTRY_BULLETS:
        return True
    text = _line_text(line_toks)
    if is_bullet_line(line_toks) and not _is_date_only_entry_line(line_toks):
        return True
    if has_date_anchor(text) and len(text.split()) >= 4:
        return True
    return False


def _is_section_heading_line(line_toks: list[dict]) -> bool:
    """EXPERIENCE section title line — not a job entry head for segmentation."""
    text = _line_text(line_toks).strip().lower()
    return text in {"experience", "work experience", "professional experience", "employment"}


def resolve_segmentation_entry_heads(tokens: list[dict], resume_id: str = "", slug: str | None = None) -> set[tuple[int, int]]:
    """Entry slice heads aligned with training evaluate.py (experienceEntryHeads first)."""
    from inference_v2.experience_phase2_divider.heads_loader import load_entry_head_lines

    mongo_heads = load_entry_head_lines(resume_id, tokens, slug) if resume_id else set()
    if mongo_heads:
        return mongo_heads
    return resolve_entry_slice_heads(tokens)


def resolve_entry_slice_heads(tokens: list[dict]) -> set[tuple[int, int]]:
    """Return line keys that begin a new experience entry for step-9 segmentation.

    Uses only lines whose first meaningful token is B-ENTRY, excluding date-only
    continuations and sub-project bullets (◦ Project …).
    """
    by_line: dict[tuple[int, int], list[dict]] = defaultdict(list)
    for t in tokens:
        key = (int(t.get("page", 0)), int(t.get("lineIndex", t.get("line_index", 0))))
        by_line[key].append(t)

    heads: set[tuple[int, int]] = set()
    for line_key, line_toks in by_line.items():
        line_toks.sort(key=lambda t: t.get("tokenIndex", t.get("token_index", 0)))
        if not _line_has_b_entry(line_toks):
            continue
        if _is_date_only_entry_line(line_toks):
            continue
        if _is_sub_bullet_line(line_toks):
            continue
        if _is_company_or_date_continuation(line_toks):
            continue
        if _is_section_heading_line(line_toks):
            continue
        first = _first_meaningful_token(line_toks)
        if first in MAIN_ENTRY_BULLETS and _first_meaningful_pred(line_toks) == "B-ENTRY":
            heads.add(line_key)
            continue
        if _first_meaningful_pred(line_toks) == "B-ENTRY":
            heads.add(line_key)
    return heads

File: web/app/inference-v2/lib/entrySliceHeads.ts
/** Primary job-entry slice heads (mirrors training-engine entry_slice_heads.py). */

const MAIN_ENTRY_BULLETS = new Set(['•', '●', '▪']);
const SUB_ENTRY_BULLETS = new Set(['◦', '∗', '*', '·']);
const STRUCTURAL = new Set(['"', "'", ',', '|', '(', ')']);
const MONTH_RE =
  /\b(jan(?:uary)?|feb(?:ruary)?|mar(?:ch)?|apr(?:il)?|may|jun(?:e)?|jul(?:y)?|aug(?:ust)?|sep(?:t(?:ember)?)?|oct(?:ober)?|nov(?:ember)?|dec(?:ember)?)/i;
const YEAR_RE = /\b(19|20)\d{2}\b/;
const APOSTROPHE_YEAR_RE = /[''\u2018\u2019\u2032]\s*(?:\d{4}|\d{2})\b/;
const DATE_END_RE = /\b(present|current|ongoing|now)\b/i;

type LineToken = {
  page: number;
  lineIndex: number;
  tokenIndex?: number;
  token?: string;
  prediction?: string;
  bioLabel?: string;
};

const predOf = (t: LineToken) => t.prediction ?? t.bioLabel ?? 'O';

const firstMeaningful = (line: LineToken[]) => {
  for (const t of line) {
    const tok = (t.token ?? '').trim();
    if (tok && !STRUCTURAL.has(tok)) return tok;
  }
  return '';
};

/** Prediction of the first non-structural token (mirrors _first_meaningful_pred). */
const firstMeaningfulPred = (line: LineToken[]): string | null => {
  for (const t of line) {
    const tok = (t.token ?? '').trim();
    if (tok && !STRUCTURAL.has(tok)) return predOf(t);
  }
  return null;
};

const SECTION_HEADING_TEXT = new Set([
  'experience',
  'work experience',
  'professional experience',
  'employment',
]);

const isSectionHeadingLine = (line: LineToken[]) =>
  SECTION_HEADING_TEXT.has(
    line.map((t) => t.token ?? '').join(' ').trim().toLowerCase(),
  );

const lineHasBEntry = (line: LineToken[]) => line.some((t) => predOf(t) === 'B-ENTRY');

const hasDateAnchor = (text: string) =>
  YEAR_RE.test(text) ||
  APOSTROPHE_YEAR_RE.test(text) ||
  DATE_END_RE.test(text) ||
  MONTH_RE.test(text);

const isDateOnlyLine = (line: LineToken[]) => {
  const text = line.map((t) => t.token ?? '').join(' ').trim();
  const first = firstMeaningful(line);
  if (!first) return false;
  if ((first === '(' || first === '|') && hasDateAnchor(text)) return true;
  return hasDateAnchor(text) && !line.some((t) => {
    const tok = (t.token ?? '').trim();
    return tok.length > 3 && !MONTH_RE.test(tok) && !YEAR_RE.test(tok) && !APOSTROPHE_YEAR_RE.test(tok);
  });
};

const isCompanyOrDateContinuation = (line: LineToken[]) => {
  const first = firstMeaningful(line);
  if (!first || MAIN_ENTRY_BULLETS.has(first)) return false;
  if (first === '(' || first === '|') return true;
  const lower = line.map((t) => t.token ?? '').join(' ').toLowerCase();
  if (hasDateAnchor(lower)) return true;
  return ['pvt', 'ltd', 'inc', 'llc', 'limited'].some((w) => lower.includes(w));
};

/** Lines that start a new experience job block for phrase segmentation. */
export const resolveEntrySliceHeadLines = (
  boundaryTokens: LineToken[],
): Set<string> => {
  const byLine = new Map<string, LineToken[]>();
  for (const t of boundaryTokens) {
    const key = `${t.page}-${t.lineIndex}`;
    if (!byLine.has(key)) byLine.set(key, []);
    byLine.get(key)!.push(t);
  }

  const heads = new Set<string>();
  for (const [key, raw] of byLine) {
    const line = [...raw].sort((a, b) => (a.tokenIndex ?? 0) - (b.tokenIndex ?? 0));
    if (!lineHasBEntry(line)) continue;
    if (isDateOnlyLine(line)) continue;
    if (SUB_ENTRY_BULLETS.has(firstMeaningful(line))) continue;
    if (isCompanyOrDateContinuation(line)) continue;
    if (isSectionHeadingLine(line)) continue;
    // A line starts an entry when its first meaningful token is predicted B-ENTRY —
    // whether that token is a bullet or a plain job title (mirrors Python
    // resolve_entry_slice_heads). Bullet-first lines were the only case handled before,
    // which dropped title-first entries like "ML Engineer …".
    if (firstMeaningfulPred(line) === 'B-ENTRY') {
      heads.add(key);
    }
  }
  return heads;
};
